lookup_estim.go

```diff
@@ -0,0 +1,302 @@
+	putProvDone atomic.Int32
+}
+
+func (dht *IpfsDHT) newEstimatorState(ctx context.Context, key string) (*estimatorState, error) {
+	// get network size and err out if there is no reasonable estimate
+	networkSize, err := dht.nsEstimator.NetworkSize()
+	if err != nil {
+		return nil, err
+	}
+
+	individualThreshold := mathext.GammaIncRegInv(float64(dht.bucketSize), 1-OptProvIndividualThresholdCertainty) / networkSize
+	setThreshold := mathext.GammaIncRegInv(float64(dht.bucketSize)/2.0+1, 1-OptProvSetThresholdStrictness) / networkSize
+	returnThreshold := int(math.Ceil(float64(dht.bucketSize) * OptProvReturnRatio))
+
+	return &estimatorState{
+		putCtx:              ctx,
+		dht:                 dht,
+		key:                 key,
+		doneChan:            make(chan struct{}, returnThreshold), // buffered channel to not miss events
+		ksKey:               ks.XORKeySpace.Key([]byte(key)),
+		networkSize:         networkSize,
+		peerStates:          map[peer.ID]addProviderRPCState{},
+		individualThreshold: individualThreshold,
+		setThreshold:        setThreshold,
+		returnThreshold:     returnThreshold,
+		putProvDone:         atomic.Int32{},
+	}, nil
+}
+
+func (dht *IpfsDHT) GetAndProvideToClosestPeers(outerCtx context.Context, key string) error {
+	if key == "" {
+		return fmt.Errorf("can't lookup empty key")
+	}
+
+	// initialize new context for all putProvider operations.
+	// We don't want to give the outer context to the put operations as we return early before all
+	// put operations have finished to avoid the long tail of the latency distribution. If we
+	// provided the outer context the put operations may be cancelled depending on what happens
+	// with the context on the user side.
+	putCtx, putCtxCancel := context.WithTimeout(context.Background(), time.Minute)
+
+	es, err := dht.newEstimatorState(putCtx, key)
+	if err != nil {
+		putCtxCancel()
+		return err
+	}
+
+	// initialize context that finishes when this function returns
+	innerCtx, innerCtxCancel := context.WithCancel(outerCtx)
+	defer innerCtxCancel()
+
+	go func() {
+		select {
+		case <-outerCtx.Done():
+			// If the outer context gets cancelled while we're still in this function. We stop all
+			// pending put operations.
+			putCtxCancel()
+		case <-innerCtx.Done():
+			// We have returned from this function. Ignore cancellations of the outer context and continue
+			// with the remaining put operations.
+		}
+	}()
+
+	lookupRes, err := dht.runLookupWithFollowup(outerCtx, key, dht.pmGetClosestPeers(key), es.stopFn)
+	if err != nil {
+		return err
+	}
+
+	// Store the provider records with all of the closest peers
+	// we haven't already contacted.
+	es.peerStatesLk.Lock()
+	for _, p := range lookupRes.peers {
+		if _, found := es.peerStates[p]; found {
+			continue
+		}
+
+		go es.putProviderRecord(p)
+		es.peerStates[p] = Sent
+	}
+	es.peerStatesLk.Unlock()
+
+	// wait until a threshold number of RPCs have completed
+	es.waitForRPCs()
```

Comment:
What happen if less than some minimum of puts fails ?
Let's say all of the puts fail, I expect to return an error the consumer (the network is probably broken), I would expect `es.waitForRPCs` to synchronise that and have it return an `error`. (this doesn't have to be `waitForRPCs` precisely)

Reply:
This would involve quite a bit of work to implement this behaviour.

Reply:
My 2c: I think it's fine to stay consistent with the classic approach and treat this as a separate issue.